### PR TITLE
[Experimental]: First version of pages list in site editor

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
This is the first version of pages list in site editor with related issues:  https://github.com/WordPress/gutenberg/issues/53233, https://github.com/WordPress/gutenberg/issues/50430.

The underlying APIs will evolve, but we need to get something in, to enable parallel work. This PR adds a new pages list in the site editor and to test it, you'll need to **enable the `New admin views` GB experiment**.

For now it uses  [TanStack table](https://github.com/tanstack/table), but this is an internal implementation detail and could also change. Part of the code has been extracted/polished a bit from https://github.com/WordPress/gutenberg/pull/53906.
where the approach was to create helper components that could be composed into creating different views.

This PR changes the direction and aims to use a high level component(`DataViews`) that will provide the APIs for core and extenders for all views and functionality for any entity. 

Nothing is exported right now and no public APIs have been introduced.

## Testing Instructions
1. Enable the experiment
2. Go to site editor and navigate to `Pages->Manage all pages` .
3. Play around with the new functionality like pagination, sorting, etc..

https://github.com/WordPress/gutenberg/assets/16275880/56f061ad-1ed2-4447-bda3-635fe7f90f98

### CODE REVIEW

File: packages/edit-site/src/components/dataviews/context.js
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/context.js]: What's the idea of this context, can you expand on it a bit? Is it something internal so we avoid prop drilling or something?

File: packages/edit-site/src/components/page-pages/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/page-pages/index.js:38]: Don't we have an entity for that?

**Review by youknowriad — COMMENTED** [review]
Since this is an experiment, and as discussed directly with Nik. I'm not going to deeply review the APIs here. I think the main goal of this PR should be to ship a starting point for us so we can iterate on this in parallel both in terms of API, design and behavior. 

I think what we need is a breakdown of these tasks and logging them into a related overview issue which should help developers that want to engage with these efforts pick up tasks.

File: packages/edit-site/src/components/page-pages/index.js
Comment (ntsekouras) [reply packages/edit-site/src/components/page-pages/index.js:38]: No. We could add one in a follow up maybe?

Comment (youknowriad) [reply packages/edit-site/src/components/page-pages/index.js:38]: Definitely

File: packages/edit-site/src/components/dataviews/context.js
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/context.js]: For this PR yes, but I had it in the other PR for extenders and it would be exported eventually. We can remove it and pass the table instance down where is needed. Do you think I should update it in this PR?

Comment (youknowriad) [reply packages/edit-site/src/components/dataviews/context.js]: I think it's fine but yeah I don't think we should expose that context.

File: packages/edit-site/src/components/sidebar-navigation-screen-pages-list/index.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar-navigation-screen-pages-list/index.js]: Do we need this to be a separate file/component?

Comment (ntsekouras) [reply packages/edit-site/src/components/sidebar-navigation-screen-pages-list/index.js]: Nope. Will remove it.

File: packages/edit-site/src/components/dataviews/context.js
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/context.js]: I removed it.

**Comment by jameskoster** [comment]
Nice work on this so far. Here are a couple of design adjustments:

For sorting, could we utilise a popover instead of auto-sorting on click. IE:

Screenshot 2023-10-02 at 14 35 01

This is a bit easier to grok, and will better set us up for expanding options in the future, e.g:

Screenshot 2023-10-02 at 14 37 00

In fact, I suppose that 'Hide' option could be added to relevant columns now too.

Sorting options can live in the view options also, e.g.:

Screenshot 2023-10-02 at 14 55 47


---

Could the pagination row be pinned to the bottom of the frame? That will avoid it jumping around during filtering etc.

---

Actions can be displayed permanently as icon buttons. I'd suggest starting with "View" (opens the page on the frontend in a new tab) and "Move to trash". These should both have tooltips.

---

It would be nice to include the featured image next to the page title, when one is set. 

Here's a mockup capturing these details:

pages

**Comment by ntsekouras** [comment]
@jameskoster this PR could be considered just a starting point and many things could be improved in follow ups where more people could work in parallel. Having said that I pushed a commit about:

>Sorting options can live in the view options also, e.g.

This works but I couldn't make it to match the design to open a third nested submenu on the left. I've checked radix docs about it a bit, but couldn't fix it. --cc @ciampo .

> Could the pagination row be pinned to the bottom of the frame?

What we should do is have a loader(maybe the progress bar) inside the table and update the new data when they arrive. I don't think we can avoid jumping because the returned rows based on filtering could be fewer/more than the current ones in view.


>It would be nice to include the featured image next to the page title, when one is set.

Let's add more fields in follow ups that we will also need to start forming the API we'll end up with.

**Comment by ciampo** [comment]
> This works but I couldn't make it to match the design to open a third nested submenu on the left. I've checked radix docs about it a bit, but couldn't fix it. --cc @ciampo .

I don't think that's really possible, because I believe that nested menu always open to the right (in LTR environments), and only open to the left when there isn't space available on the right. So in this case, the third nested submenu has the space to open to the right and it does so. My hunch is that this behaviour can't be changed

Also, a heads-up that the experimental `DropdownMenuV2` likely won't stick around too much in its currently iterations (see https://github.com/WordPress/gutenberg/issues/50459)

**Comment by ntsekouras** [comment]
>Also, a heads-up that the experimental DropdownMenuV2 likely won't stick around too much in its currently iterations (see https://github.com/WordPress/gutenberg/issues/50459)

Are you considering just changing the internal implementation or the whole structure/components etc..?

File: packages/edit-site/src/components/page-pages/index.js
Comment (oandregal) [reply packages/edit-site/src/components/page-pages/index.js:38]: I'll work on this.

File: packages/edit-site/src/components/dataviews/list-view.js
Comment (andrewserong) [inline packages/edit-site/src/components/dataviews/list-view.js:39]: This could be a bit of a question for design across some of this work, but is it possible for us to come up with a more unique name for the list view of table data, so that we don't confuse it with the `ListView` component in the sidebar of the post and site editors? Maybe prefixing with `Table` would be enough — `TableListView`?

**Review by andrewserong — COMMENTED** [review]
This is exciting! I haven't given the code a detailed look, but just added a tiny comment about `ListView` naming and whether we can make it more unique to distinguish it from the list view sidebar.

One question in terms of design (and given this is an experiment, it can totally be looked at separately), but now that there's a view of table data that updates in a real-time way, it could be good for the pagination area to be sticky-ied to the bottom of the view, as currently it jumps up based on the number of results:



In any case, as Riad mentioned (https://github.com/WordPress/gutenberg/pull/54966#pullrequestreview-1652577824), I'm sure this sort of polishing can be picked up as separate tasks after the initial experiment has been merged 👍

Looking forward to seeing how all this evolves!

File: packages/edit-site/src/components/dataviews/list-view.js
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/list-view.js:39]: Yeah, we could rename, but this component will not be exposed and will only be used internally by the high level `DataViews` component.

Comment (youknowriad) [reply packages/edit-site/src/components/dataviews/list-view.js:39]: this could be named "TableView" maybe to avoid confusion, we'll have things like "GridView", "KanbanView"... later

**Comment by ciampo** [comment]
> > Also, a heads-up that the experimental DropdownMenuV2 likely won't stick around too much in its currently iterations (see #50459)
> 
> Are you considering just changing the internal implementation or the whole structure/components etc..?

Probably a mix of both, mostly dictated by some differences between `Ariakit` and `radix-ui`. But the overall idea of composing menus and submenus via `children` should stay — what could change are a few props and component names.

**Review by youknowriad — APPROVED** [review]
So this is clearly unfinished and unpolished but I'm approving so that we can have a starting point that allows us to iterate on in parallel.

File: packages/edit-site/src/components/page-pages/index.js
Comment (oandregal) [reply packages/edit-site/src/components/page-pages/index.js:38]: https://github.com/WordPress/gutenberg/pull/55050

File: packages/edit-site/src/components/page-pages/index.js
Comment (oandregal) [inline packages/edit-site/src/components/page-pages/index.js:138]: Created a follow-up at https://github.com/WordPress/gutenberg/pull/55196 It seems we should use `accessorFn` instead of `cell` to render this value.

**Comment by getdave** [comment]
✅ I updated the [PHP Sync Tracking Issue](https://github.com/WordPress/gutenberg/issues/57959) to note this PR does not require a backport.

**Pull request merged**
